Upgrade to uglify-js 3.0.4

alexlamsl opened the pull request:
Fixes #471, fixes #472, fixes #442.

XhmikosR commented on docs/uglify-options.md:
```diff
@@ -7,113 +7,94 @@ This task primarily delegates to [UglifyJS2][], so please consider the [UglifyJS
 
 
 ## mangle
-Type: `Boolean` `Object`  
```
@alexlamsl: can you put back the trailing spaces in markdown files? They are used for line breaks.

alexlamsl replied:
Sorry about that - my editor was configured to automatically remove them.

I'll go through the changes and restore them.

XhmikosR replied:
No worries. BTW, you can skip README.md in your patch; it will get updated automatically with the next version bump.

alexlamsl replied:
Ah, that explains the repetition from `docs/uglify-options.md` etc.

I'll revert this file then.

XhmikosR commented:
@alexlamsl: can you sign the CLA agreement too?

As for the changes themselves, they are too many :/ I'll try to review most of them but it's my understanding you removed our custom options?

XhmikosR commented on tasks/uglify.js:
```diff
@@ -48,17 +49,11 @@ module.exports = function(grunt) {
       banner: '',
       footer: '',
-      compress: {
-        warnings: false
```
Are warning off by default now?

alexlamsl replied:
Yes.

XhmikosR commented on docs/uglify-options.md:
```diff
@@ -7,113 +7,94 @@ This task primarily delegates to [UglifyJS2][], so please consider the [UglifyJS
+Type: `Boolean` `Object`
 Default: `false`
 
 Turns on beautification of the generated source code. An `Object` will be merged and passed with the options sent to `UglifyJS.OutputStream()`. [View all options here](https://github.com/mishoo/UglifyJS2#beautifier-options)
```
This seems invalid now? There might be more instances in docs.

alexlamsl replied:
I've retained this shorthand here https://github.com/gruntjs/grunt-contrib-uglify/pull/472/files#diff-afaf7edd36b493b2d939b60518620ec6R60

XhmikosR replied:
But it's my understanding that `OutputStream()` is no longer available. That is where I'm referring to :)

alexlamsl replied:
Ugh - of course! :sweat_smile: 

Will remove that sentence.

alexlamsl commented:
> As for the changes themselves, they are too many :/ I'll try to review most of them but it's my understanding you removed our custom options?

Apologies for the volume of changes - switching over to `minify()` should hopefully be more maintainable from here on out.

I've removed custom options that are direct case-renaming of `minify()` options like `ASCIIOnly`. Things that have special meanings within this project like `sourceMapIn` and `sourceMapName` are kept, just updated to set the new options within `sourceMap`.

XhmikosR commented:
Ah, I see. I'm all for removing custom options. I just want to handle this the easiest and clearest way for people who upgrade. Perhaps we should make a paragraph about breaking changes from our side and also add a link to the upstream breaking changes if there's one?

alexlamsl commented:
> Perhaps we should make a paragraph about breaking changes from our side and also add a link to the upstream breaking changes if there's one?

I can tabulate the list of deprecated custom options and their corresponding names in `minify()`. Which markdown files under `docs/` should I be adding that to?

XhmikosR commented:
`uglify-options.md` seems like the best place to add them.

Thanks for all your help!

PS. I don't watch all the repos so for future issues/PRs please CC me :)

XhmikosR commented:
One last thing: how about the test files you removed (enclose.js, exportAll.js)? Are those options obsolete or something?

alexlamsl commented:
> One last thing: how about the test files you removed (enclose.js, exportAll.js)? Are those options obsolete or something?

Indeed - those two features have been removed from 3.x

alexlamsl commented:
> can you sign the CLA agreement too?

OT: I guess I'm one of those oddballs that would actually try and read the licence, though with `[JS Foundation IP Policy]({{ site.url }}{{ site.baseurl }}/pdf/ip-policy.pdf)` I'm not sure how much further I can get into it :rofl:

XhmikosR commented:
> OT: I guess I'm one of those oddballs that would actually try and read the licence, though with `[JS Foundation IP Policy]({{ site.url }}{{ site.baseurl }}/pdf/ip-policy.pdf)` I'm not sure how much further I can get into it 🤣

Sorry about that, @vladikoff might know how to fix it :/

XhmikosR commented:
So a couple more questions. Do you plan to support v2.x? If so we should
make a branch here too.

Secondly will uglify throw an error if an invalid option is passed ? Say
one passes an option from v2 which has been deprecated.

On May 9, 2017 18:27, "Alex Lam S.L." <notifications@github.com> wrote:

can you sign the CLA agreement too?

OT: I guess I'm one of those oddballs that would actually try and read the
licence, though with [JS Foundation IP Policy]({{ site.url }}{{
site.baseurl }}/pdf/ip-policy.pdf) I'm not sure how much further I can get
into it 🤣

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/gruntjs/grunt-contrib-uglify/pull/472#issuecomment-300200936>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAVVtQrN47OoSOATDtJd7EOHE0BliLV5ks5r4IXLgaJpZM4NVQ5V>
.

alexlamsl commented:
> So a couple more questions. Do you plan to support v2.x? If so we should
make a branch here too.

We do have a `2.x` branch for backports of bug fixes only, at least for a while. That means you'll only get `2.8.x`, so I think the current version of `grunt-contrib-uglify` should cover that without any modifications.

Obviously if you are planning on pushing fixes to `2.x` of `grunt-contrib-uglify` as well, then making a branch makes sense.

> Secondly will uglify throw an error if an invalid option is passed ? Say
one passes an option from v2 which has been deprecated.

`UglifyJS.minify()` will throw if invalid options are passed to it, which was the case in v2 as well. However, in `tasks/lib/uglify.js` I take the approach of constructing an options object from known keys and pass that onward to `minify()` instead. So any (top-level) invalid options from `grunt-contrib-uglify` would simply be ignored.

XhmikosR commented:
If you think there's no need to keep 2 separate branches then we don't need to. I'm not sure why you keep a branch for 2.x in the first place :)

As for the unknown options, I guess we can keep the same behavior. Personally, I don't like silent failures but that is for another time.

@vladikoff: please address https://github.com/gruntjs/grunt-contrib-uglify/issues/471#issuecomment-300078551 and https://github.com/gruntjs/grunt-contrib-uglify/pull/472#issuecomment-300200936

vladikoff commented:
If there is ever a security issue for 2.x we probably would have to make a release. But we can make that branch if needed when that happens.

vladikoff commented:
@XhmikosR should I do a final review of this?

XhmikosR commented:
@vladikoff: I mostly meant https://github.com/gruntjs/grunt-contrib-uglify/pull/472#issuecomment-300200936 about the CLA :)

alexlamsl commented:
> If you think there's no need to keep 2 separate branches then we don't need to. I'm not sure why you keep a branch for 2.x in the first place :)

In case my screw-up receives knighthood. :sweat_smile: 

> As for the unknown options, I guess we can keep the same behavior. Personally, I don't like silent failures but that is for another time.

I share the same opinion, but am unfamiliar with `grunt` to know what the norm is around here.

vladikoff commented:
CLA is at https://js.foundation/CLA

where is the broken link?

XhmikosR commented:
I believe most plugins just do this. @vladikoff: what is your opinion about unknown options handling? Currently we don't show anything and wrong switches are being silently ignored. Note that this change should be made in most if not all plugins so it's for the future.

About this PR, we can all do a final review and squash and merge it when we are all done. I should have some time in a few hours. Also, I suppose the new release should be v3.0.0 and I vote to keep a 2.x branch just in case we need it in the future.

alexlamsl commented:
@vladikoff on the signing page https://cla.js.foundation/gruntjs/grunt-contrib-uglify?pullRequest=472

JS Foundation projects, including code contributions and all related materials, are each released under the terms of the individual licenses as noted in the project's repository, or, if no separate license is specified, under the terms of the Apache v2.0 license (the "Applicable License"), with the exception of project documentation and sample code, in accordance with the **[JS Foundation IP Policy]({{ site.url }}{{ site.baseurl }}/pdf/ip-policy.pdf)**.

alexlamsl commented:
@XhmikosR +1 for `v2.x` branch before merging this PR - it's free anyway :wink:

vladikoff commented:
> @vladikoff on the signing page https://cla.js.foundation/gruntjs/grunt-contrib-uglify?pullRequest=472 JS Foundation projects, including code contributions and all related materials, are each released under the terms of the individual licenses as noted in the project's repository, or, if no separate license is specified, under the terms of the Apache v2.0 license (the "Applicable License"), with the exception of project documentation and sample code, in accordance with the [JS Foundation IP Policy]({{ site.url }}{{ site.baseurl }}/pdf/ip-policy.pdf).

@dmethvin ^ what do we do about that?

dmethvin commented:
@vladikoff Hmmm, not sure why that link is broken but it's supposed to point to https://js.foundation/pdf/ip-policy.pdf which has also gone missing. I'll page @kborchers who can probably clear things up. @alexlamsl thanks for finding this!

XhmikosR commented on tasks/uglify.js:
```diff
@@ -19,7 +20,7 @@ function relativePath(file1, file2) {
 
   if (file1Dirname !== file2Dirname) {
-    return path.relative(file1Dirname, file2Dirname) + path.sep;
+    return path.relative(file1Dirname, file2Dirname);
```
@alexlamsl; what is the reasoning behind this change?

alexlamsl replied:
`relativePath()` is only used once, down here:
https://github.com/gruntjs/grunt-contrib-uglify/pull/472/files#diff-0c8f71b28b3ffffd4d019110b7446b88L139

Since I'm now using `path.join()` instead of string concatentation down below, that trailing (back)slash is no longer required.

XhmikosR replied:
Thanks for the explanation.

We should be using `path.posix` methods when we drop node.js 0.10 support but that's for the future.

XhmikosR commented on test/uglify_test.js:
```diff
@@ -21,20 +21,20 @@ exports.contrib_uglify = {
       'sourcemap_basic.js.map',
       'sourcemap_customDir.js',
       'sourcemap_customName.js',
+      'source_map_custom_name',
```
Is this right? It's missing the file extension.

alexlamsl replied:
The test case has been like that before this PR:
https://github.com/gruntjs/grunt-contrib-uglify/blob/master/Gruntfile.js#L176
https://github.com/gruntjs/grunt-contrib-uglify/blob/master/test/fixtures/expected/source_map_custom_name

But for some reasons this generated map file isn't being compared at all, so I thought I'd add it to the list here.

XhmikosR replied:
Hmm, I see now. Indeed. Cool, just leave it like that.

XhmikosR commented:
@alexlamsl: thanks for all the help! I'm gonna release v3.0.0 now.

vladikoff commented:
@XhmikosR Amazing.....!!!

kzc commented:
@alexlamsl Any interest in porting `uglifyjs-webpack-plugin` to use `uglify-es` or `uglify-js@3`?

https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/32

It's just one source file:

https://github.com/webpack-contrib/uglifyjs-webpack-plugin/blob/master/src/index.js

alexlamsl commented:
@kzc I believe they have a fork that works & is just waiting to merge upstream.

Combination of an ES6 codebase, extra dependencies & the fact that I've never used `webpack` before means I'm not the best candidate out there... 😅
